### PR TITLE
Arctic MoE

### PR DESCRIPTION
**Pull request by sfc-gh-sbekman**
# Arctic MoE (AMoE)

## Description

This is a new implementation that re-uses the ZeRO-2 integration of DeepSpeed-MoE from 4 years ago, but everything else is new and modernized. We use gpt-oss approach of packing all experts into just 2 tensors (1. gate+up 2. down) which allows for a much more efficient kernel compute (unlike HF implementation that loops over experts with low compute density).

## Current status

- [x] Expert Parallelism works
- [x] Shared Expert works
- [x] Qwen3-Moe works including exporting back to original  plus save/resume support- the loss is correct. Qwen3-30B fits into 8xh100 with EP=8 - could probably do some sl=80k may be even longer
- [x] Qwen3-Next works including exporting back to original plus save/resume support
- [x] Setting up dense vs moe equivalent benchmarks at the moment for Qwen3-30B on gpu (no offload) we are about 3.3x slower than the dense equivalent when adapting the seqlen to match tokens per mlp params
- [x] Reza is working on efficient kernels
- [x] Unit tests are there
- [x] Added support for linear gated delta net attention in the flop counter
- [x] Added by component time and token profiler activated via Deepspeed flops profiler - so now we can tell which ops are slow while checking the proportion of the tokens moved per rank

## TODO before others can use it practically

- some tweaks remaining wrt exporting back to HF format
- performance optimizations.

## Trying it out

You can run the benchmark recipes [here](https://github.com/snowflakedb/ArcticTraining/blob/stas/ds-moe/projects/arctic_moe/weekly-benchmark/README.md).

## Limitations

Things that I have observed that limit things:

- Currently only ZeRO Stage2 is supported since that's how the original DS-MoE was built - ideally need to integrate into ZeRO Stage3 as well - Z3 will help with models like Qwen3-moe since it has more dense than MoE weights, Qwen-Next-80B is 90% MoE params so Z3 will help very little here.  
- cpu-offload in Z2 is really slow, especially during the model loading stage, which often doesn't complete in 30min needing to extend the torch.dist 30min timeout. The per iteration performance is quite slow too since it has to update simultaneously all MoE params from 8 processes. I have observed an almost 2x slower throughput as compared to non-cpu-offload version.

### CODE REVIEW

File: arctic_training/model/moe/moe.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+import torch.nn.functional as F
+from moe_gem import MoEGEMM
```
[the remark this answers is not in the document]
Comment (sfc-gh-reyazda) [reply arctic_training/model/moe/moe.py:22]: I am finishing some TMA optimization for some triton kernels and will add them soon

[the remark this answers is not in the document]
Comment (sfc-gh-sbekman) [reply arctic_training/model/moe/moe.py:22]: thank you - in general if we share a branch - let's try to push code that works or disabled so that the branch remains functional

**Review by sfc-gh-mwyatt — APPROVED** [review]
